lib/Dialects/VectorExt/Transform/VectorMaskingUtils.cpp

```diff
@@ -0,0 +1,138 @@
+  // Compute the range of operations that will be moved within vector.predicate
+  // and the definitions within the range with external users.
+  Block &bodyBlock = loopOp.getLoopBody().front();
+  auto opsToMove =
+      llvm::make_range(bodyBlock.begin(), Block::iterator(bodyBlock.back()));
+
+  SmallVector<Value, 8> defsWithExternalUses;
+  SmallVector<Type, 8> defTypes;
+  getDefsWithExternalUses(opsToMove, defsWithExternalUses, defTypes);
+
+  // Initialize a builder to insert vector.predicate at the beginning of the
+  // loop body.
+  Location loc = loopOp.getLoc();
+  MLIRContext *context = loopOp.getContext();
+  OpBuilder builder(loopOp);
+  builder.setInsertionPointToStart(&bodyBlock);
+
+  // Generate the predicate to used by the vector.predicate operation:
+  //   %min = affine.min affine_map<(d0)[s0] -> (vecFactor, -d0 + s0)>(%iv)[%ub]
+  //   %vec_pred = vector.create_mask %min : vector<8xi1>
+  // TODO: This is probably not the most efficient way to generate the vector
+  // predicate. Consider using the vector IV.
+  AffineExpr i, j;
+  bindDims(context, i, j);
+  SmallVector<AffineMap, 4> maps = AffineMap::inferFromExprList(
+      ExprList{{builder.getAffineConstantExpr(vecFactor), i - j}});
+  auto minOp = builder.create<AffineMinOp>(
+      loc, loopOp.step()[0].getType(), maps[0],
+      ValueRange{loopOp.upperBound()[0], loopOp.getInductionVars()[0]});
+
+  auto maskType = VectorType::get({vecFactor}, builder.getI1Type());
+  auto predicate =
+      builder.create<CreateMaskOp>(loc, maskType, ValueRange{minOp});
+
+  // Generate the vector.predicate operation and move the range [`startOp`,
+  // `endOp`] of operations within its truePredicateRegion. We have to rewire
+  // the def-use chain for those definitions within the range that have external
+  // uses. Those uses are rewired to the results of the vector.predicate.
+  auto vecPredOp = builder.create<PredicateOp>(loc, defTypes, predicate);
+  assert(defsWithExternalUses.size() == vecPredOp.getNumResults() &&
+         "Expected same size");
+  for (auto &en : llvm::enumerate(defsWithExternalUses))
+    en.value().replaceAllUsesWith(vecPredOp.getResult(en.index()));
+
+  Operation *truePredTerminator =
+      &vecPredOp.truePredicateRegion().front().back();
+  moveOperationsBefore(opsToMove, truePredTerminator);
+
+  // The existing terminator of TruePredicateRegion doesn't yield any value.
+  // Replace it with a new terminator that returns the definitions with external
+  // uses that we just moved.
+  if (vecPredOp.getNumResults() > 0) {
+    builder.setInsertionPoint(truePredTerminator);
+    builder.create<vector_ext::YieldOp>(loc, defsWithExternalUses);
+    truePredTerminator->erase();
+  }
+}
+
+/// Utility that predicates a tiled loop with a vector.predicate operation. The
+/// vectorization factor used for predication is assumed to be the step of the
+/// tiled loop.
+LogicalResult mlir::vector_ext::predicateTiledLoop(TiledLoopOp loopOp) {
```

Comment:
Same comment re anchoring on `TiledLoopOp`.
Actually, can the general impl. of this be achieved by outlining inside an scf.execute op + region ?

Reply:
It should be better now.

Reply:
Hmm I still see this anchoring on `TiledLoopOp`.
Can this be implemented on Operation* with some helper function that would later turn into an interface on the right ops?

Reply:
Look at `predicateOp` above. That's the generic implementation on `Operation*`. `predicateTiledLoop` is just extracting the information needed from the TileLoopOp and the invoking the generic implementation (line 147).